### ISSUE TITLE
[RFC] Faster CPU sampling through fused sampling+compaction

### ISSUE THREAD

**Issue by hesham-mostafa** [dmlc/dgl#5328]
## 🚀 Feature

This issue is for discussing fused sampling kernels that sample and compact the sampled graph in one step, and directly return a block in CSC(compressed sparse column) format. This completely bypasses the expensive ``to_block`` function. Experimental implementation is [here](https://github.com/hesham-mostafa/dgl/tree/sampling_optim)

## Motivation

In sampling-based GNN training on CPUs, we observe that graph sampling takes up a significant portion of the epoch time. Graph sampling often ends up taking more time than the actual GNN forward+backward time. The current sampling pipeline in DGL is illustrated in the figure below:

![dgl_sampling_problem svg 2023_02_19_09_57_03 0](https://user-images.githubusercontent.com/16276103/219972964-58986534-d634-4077-af4c-f8433dfe768b.svg)

In the current sampling pipeline, creating a graph sample that can be passed to a GNN is a 2-step process:
1. ``dgl.sampling.sample_neighbors`` or similar functions takes the seed nodes and sample some neighbors of the seed nodes. They return a copy of the original graph in COO(coordinate) format that only contains the sampled edges.
2. ``dgl.to_block`` compacts the sampled graph to create a bipartite graph. ``dgl.to_block`` reindexes the nodes so that the indices of the source nodes and the target nodes are consecutive and zero-based. It is common to include the target nodes at the beginning of the target nodes (``include_dst_in_src = True``).

``dgl.to_block`` takes up the majority of the sampling time as it involves expensive hashing operations that are needed to compact the graph. Compaction is the process of taking the original node indices and renumbering them so that they are contiguous and start from zero. We argue that several of the steps needed to compact the sampled graph can be very cheaply implemented if they are integrated with the actual graph sampling steps. The graph sampling step would take longer, but the extra time is much less than the time needed by the ``dgl.to_block`` function.

## Alternatives

There is currently no alternatives in DGL to the 2-step sampling process outlined above. 

## Pitch

Implement C++ functions that sample and compact the sampled graph on the fly, and directly returns a compacted bipartite graph as a CSR matrix. This graph can be directly passed to a GNN, bypassing the ``dgl.to_block`` function.

## Additional context

The actual sampling of the graph neighbors is mostly done using functions such as ``CSRRowWisePick`` defined [here](https://github.com/dmlc/dgl/blob/master/src/array/cpu/rowwise_pick.h). These functions return COO matrices that contain the uncompacted topology of the sampled graph using the original node indices. We propose to define a parallel set of functions that sample the graph, and compact it on the fly. These functions will return compacted graphs in the form of CSR matrices that can be directly passed to GNNs. This bypasses the ``dgl.to_block`` function. The reasons we want to fuse compaction to sampling are:
1. Compacting the target/seed nodes is trivial during sampling. We assume the seed nodes are unique, and the ith seed node is simply mapped to target node i.
2. We have immediate access to the number of source nodes for each target node (the fan-in), as we get this directly from the ``NumPicksFn`` before we start sampling. We can thus construct the ``indptr`` component of our CSR matrix practically for free.  

We still have to compact the source nodes. In our current implementation, for compacting the source nodes, we use a naive O(N) hashing where N is the number of nodes in the graph. It is very fast but takes up more memory than more elaborate techniques. A sample implementation of a fused ``CSRRowWisePick`` is [here](https://github.com/hesham-mostafa/dgl/blob/sampling_optim/src/array/cpu/rowwise_pick.h#L535).   

## API and usage scenarios

We propose to add a new flag to ``dgl.sampling.sample_neighbors``. If this flag is set, ``sample_neighbors`` will dispatch to a C++ function that does combined sampling and compaction. The flag will be ``False`` by default.

The current ``dgl.dataloading.NeighborSampler`` runs the 2-step process of sampling the graph, followed by compaction using ``dgl.to_block``. We propose to add a new sampler ``dgl.dataloading.NeighborSamplerFused`` that will call the ``dgl.sampling.sample_neighbors`` so that it dispatches to the combined sampling and compaction kernels.  ``dgl.dataloading.NeighborSamplerFused`` thus does not call ``dgl.to_block``. 

## Experimental results

We implemented the fused sampling+compaction [here](https://github.com/hesham-mostafa/dgl/tree/sampling_optim). We measured performance on a 36-core Icelake processor with 256GB of RAM. We downloaded the stochastic training of GNNs tutorial code from [here](https://docs.dgl.ai/tutorials/large/L1_large_node_classification.html). We modified it by adding these lines:

```python

fused = True #Enable fused sampling+compaction 
if fused:
    sampler = dgl.dataloading.NeighborSamplerFused([15, 10, 5])
else:
    sampler = dgl.dataloading.NeighborSampler([15, 10, 5])

```

We tested on ``ogbn-arxiv``, ``ogbn-products``, and ``ogbn-papers100M``. We used a 3-layer GraphSage model with hidden layer size of 128, batch size of 1024, and sampling fanouts of [15,10,5]. We tested with 0,2, and 4 DataLoader workers and enabled the dataloader's CPU affinity for dataloader workers larger than 0. The 'sampling time' is the time spent waiting on the dataloader. This will only accurately reflect the time needed to sample and compact the graph when the number of workers is zero. The 'gnn time' is the time spent in the GNN's forward+backward pass.

![arxiv_sampling](https://user-images.githubusercontent.com/16276103/219973370-52c54f28-4384-47ea-ba16-ba8e604ee8e3.png)

![products_sampling](https://user-images.githubusercontent.com/16276103/219973382-416a48c9-3a1c-4699-a465-c1d60ff6008e.png)

![papers100_sampling](https://user-images.githubusercontent.com/16276103/219973396-05a73452-3a83-4328-80bb-757b1303ce7c.png)

The results indicate that the fused sampling+compaction kernel runs significantly faster than the vanilla sampling approach that uses ``dgl.to_block`` (up to ~2.5X faster). As we increase dataloader workers, the parallelized sampling gradually hides the sampling latency, but the fused sampling+compaction kernels still lead to significantly faster epoch times.

**Comment by peizhou001** [comment]
Hi @hesham-mostafa, thanks for the detailed docs and tests. Could you elaborate more so we can better understand it?

1. How much performance gains under different batch size?
2. Is the speed up comes from the data locality and reduction of graph format?

As we have did some improvements for the to_block CPU, you can refer latest master branch and use it for comparison.

Then I'm not sure if the limitation exists in below 3 places:
1. In Dist-DGL, `seed nodes = local nodes + remote nodes`, can fused kernel feasible here?
2. Is there a risk some sampling algorithms doesn't support **fused**?
3. Is it supported in heterogeneous graph?

**Comment by hesham-mostafa** [comment]
Hi @peizhou001  . Sorry for the delayed response. We were doing some further work and optimizations on our fused sampling kernels in the last month. It is significantly faster now. In the following plots, we always compare to the recent optimized ``to_block`` in DGL's master branch (the version that uses concurrent hashes). Tests were on a 2-socket Icelake Xeon machines (2*36=72 cores)
1. The performance gains are significant at different batch sizes. The plots show the sampling time per epoch for different batch sizes with `num_workers=0` in the dataloader. The GNN forward/backward time is virtually the same for our fused sampling kernels and the vanilla sampling pipeline (which uses ``sample_neighbors`` followed by ``to_block`` and `test`)

<img width="396" alt="image" src="https://user-images.githubusercontent.com/16276103/229229070-f7131700-df40-47d3-b8b9-459ff6223d01.png">
<img width="415" alt="image" src="https://user-images.githubusercontent.com/16276103/229229081-b1fbd935-2106-4803-9eaa-22ffbe3989df.png">

2. There are several reasons for the speed-up:
     1. We do not use hashes, but use a mapping array whose number of elements is equal to the number of nodes in the graph. The mapping array allows us to very quickly map the global node ids to local node ids (local to the sampled graph) by constructing the mapping array in a concurrent manner `test` so that `mapping[global_node_id]= local_node_id`. The extra memory cost of the mapping array vs. traditional hashing is not significant when training on CPUs. 
     2. Another reason for the speedup is that by fusing the ``to_block`` functionality to the sampling routine, we immediately have access to the fan-in of the DST nodes and do not need to recalculate the fan-ins to construct the CSC matrix. 
     3. We also get rid of a lot of intermediate tensors (for example the intermediate COO matrix  with global indices produced by the sampling routines)
3. Our fused sampling kernel will not work as is in Dist-DGL. Our kernel fuses together the functionality of ``dgl.sampling.sample_neighbors`` and ``dgl.to_block``. In Dist-DGL, however, sampling proceeds through multiple calls to ``dgl.sampling.sample_neighbors`` to sample from every worker the neighbors of the seed nodes owned by that worker. The resulting sampled graphs (with global indices) are then merged and passed to ``to_block``. We could think of supporting this scenario in the future.
5. We added support for heterogeneous graphs [here](https://github.com/hesham-mostafa/dgl/tree/sampling_optim_heterograph ). Initial results on ogbn-mag for different batch size also indicate that the fused sampling is 2-3X faster than the current DGL's sampling pipeline (that uses concurrent hashes)

**Comment by jermainewang** [comment]
Thanks for the detailed RFC and explanation. This is in general a great effort to accelerate a critical part of the end-to-end GNN training, so we'd like to see this being integrated into DGL. According to the offline discussion, there are two options moving forward:

**Option.1**: Add this as a new operator to the `dgl.sampling` namespace. Reimplement `dgl.dataloading.NeighborSampler` ([this part](https://github.com/dmlc/dgl/blob/cc5c2379650bfda06bf331ad42d565962032b7af/python/dgl/dataloading/neighbor_sampler.py#L141-L160)) to invoke fused operator. This requires the new fused implementation to be compatible with the existing interface of `dgl.sampling.sample_neighbors` such as supporting `exclude_edges`. This option is more user-friendly because the optimization requires no user-side code changes.

**Option.2**: Add a new sampler object to the `dgl.dataloading` namespace as an alternative to the original `dgl.dataloading.NeighborSampler`. The two samplers may not need to share the same interface. However, the option is less user-friendly since users need to learn the new API.

We should try to target Option.1 first and see if there are any blockers. Option.2 is generally a temporary workaround.

The other conclusion is that this effort can be done independently with the existing effort (called GraphBolt) to refactor `dgl.dataloading` and `dgl.sampling`. Feel free to contribute it directly to current DGL codebase and we will handle the migration later.

cc @frozenbugs

**Comment by jermainewang** [comment]
In terms of benchmarks, the datasets you selected (ogbn-products and ogbn-papers100M) are good. We'd like to further see the performance under different sampling configurations:

* Different fanout values. Typical cases are: 10, 15, 30
* Different number of hops. Typical cases are: 2, 3
* Getting the entire neighborhood instead of sampling. This is needed by some evaluation task. In the original `dgl.dataloading.NeighborSampler`, this can be configured by passing fanout = -1.
* Larger batch sizes, e.g., 10K, which is possible for some link prediction task.

cc @Rhett-Ying
